### PR TITLE
Dynamic call scripts part 1 - fill in representative info

### PR DESCRIPTION
**Pull request by samhavens**
This is in response to #85, though it may not cover all of the desired behavior.

*Summary:*

- Add dependency on `gender-detection` package (see discussion below)
- Change signature of `script` component to take `currentContact`
- Add templating to script component to fill in the user's current contact's information

*Example:*

Prior behavior:

> Hi, my name is [Name] and I'm a constituent from [City, State, Zip].
>
>I’m calling to tell [Senator/Rep’s Name] that our community does not support Trump’s Muslim ban, and we expect [him/her] to take an immediate stand against this unjust executive order.
>
>[Optional: add anything else you’d like here]
>
>Thank you for your time.

With this change:

> Hi, my name is [Name] and I'm a constituent from [City, State, Zip].
>
>I’m calling to tell Ms. Feinstein that our community does not support Trump’s Muslim ban, and we expect her to take an immediate stand against this unjust executive order.
>
>[Optional: add anything else you’d like here]
>
>Thank you for your time.

*Discussion:*

I was really reluctant to infer gender. I have seen some fights in other open-source projects about this exact issue. However, without doing so, the filled-in scripts just felt awkward. I feel like the point of this modification is to make the process even _easier_ for people who are uncomfortable on the phone, and giving them an awkward-sounding script wouldn't accomplish that. Also, currently, all representatives are gender binary, so that's a factor too.

So, why this particular package for inferring gender? It doesn't use `fs`, so it can be browserified. There may be other options, I examined a few, and this one seemed ok.

*Todo:*

Still need to store the user's information in localStorage and fill that in as well.

### CODE REVIEW

**Comment by nickoneill** [comment]
Thanks @samhavens! We'll communicate to the writing team to standardize on these replacement strings. 

In the meantime, something like "lindsey graham" is incorrectly identified. I expect there are only a handful of these in our dataset but we should have some way of overriding them if a user tells us they're incorrect (or if they pop to the top of the head, like Senator Graham).

**Comment by samhavens** [comment]
Good point. I was thinking, it may be easier to distinguish Senator vs
Representative?

**Comment by nickoneill** [comment]
Yeah, that might make things easier @samhavens. There should be an `area` field for each contact noting if they're in the House, Senate or Other.

File: static/js/components/script.js [hunk 1]
```diff
@@ -1,18 +1,33 @@
...
-    
+
+    const repGender = gender.detect(contact.name);
+    const titleMap = { male: 'Mr. ', female: 'Ms. ', unisex: '' };
+    const pronounMap = { male: 'him', female: 'her', unisex: 'them' };
+    const repTitle = titleMap[repGender];
+    const formalName = `${repTitle}${contact.name.split(' ').pop()}`;
+
+    const scriptFormatter = (issue) =>
+      issue.script.replace('[him/her]', pronounMap[repGender])
```
Comment (oderby) [inline static/js/components/script.js:17]: It seems it's not currently used, but why not also support [he/she] substitution?

File: static/js/components/script.js [hunk 2]
```diff
@@ -1,18 +1,33 @@
...
 
-module.exports = (state, prev, send) => {
+module.exports = (contact, state, prev, send) => {
 	  const issue = find(state.issues, ['id', state.location.params.issueid]);
 	  const currentContact = issue.contacts[state.contactIndex];
-    
+
+    const repGender = gender.detect(contact.name);
+    const titleMap = { male: 'Mr. ', female: 'Ms. ', unisex: '' };
+    const pronounMap = { male: 'him', female: 'her', unisex: 'them' };
```
Comment (oderby) [inline static/js/components/script.js:12]: style nit/question: Do we care about putting LUT like this outside the function (to line 5)?

File: static/js/components/script.js [hunk 3]
```diff
@@ -1,18 +1,33 @@
 const html = require('choo/html');
 const find = require('lodash/find');
+const gender = require('gender-detection');
 const scriptLine = require('./scriptLine.js');
 
-module.exports = (state, prev, send) => {
+module.exports = (contact, state, prev, send) => {
 	  const issue = find(state.issues, ['id', state.location.params.issueid]);
 	  const currentContact = issue.contacts[state.contactIndex];
```
Comment (oderby) [inline static/js/components/script.js:8]: seems like `contact === currentContact` (line 8 here is same as line 26 in call.js). Either delete line 8 or don't need to add the argument.

Oh, just noticed line 7 here is same as line 11 above...

And looks like we're passing contact by argument to contact.js in 30 in call.js...

Seems like we should decide how we want to handle these redundant state lookup/extraction needs. It seems we currently have two different practices: look it up once, and pass by argument downwards (the case of `contact` here and contact.js) vs. look it up independently in every component that needs it (e.g. `issue` here and call.js). My intuition is to require all components to have a standardized interface (that being `(state, prev, send)`) and providing helpers/utilities for looking up specific/common states from the state object if that becomes onerous. But, I'm new to Choo/this code base, so will happily defer to @nickoneill or others on which we should follow. But whatever we choose, we should document and be consistent.

**Review by oderby — COMMENTED** [review]
Do we care about using the appropriate title for married women? (Ms. vs Mrs.)? Should we just avoid it by using their other title of 'Senator' or 'Representative'? I guess 'Representative' is long... but could just shorten to 'Rep.'. There's also consideration for other people outside of Senators and Representatives (President, Judge, Governor)...if we have any ambitions of calling those folks, we might want to consider a more robust treatment of titles (and how they're indicated in scripts)

**Comment by nickoneill** [comment]
@samhavens let me know if you want to make the proposed changes, otherwise I'll hand it off to someone else.

**Comment by samhavens** [comment]
I am happy to. Sorry, I am new to the project and don't know how long to wait for consensus.

Want me to encapsulate the gender info more, add he/she map, but change the actual script replacements to be to Rep. and Senator? Also, what's the decision on changing the signature of the exported function?

File: static/js/components/script.js (hunk 3 above)
Comment (nickoneill) [reply static/js/components/script.js:8]: Yep, I think this is just a case of me trying it different ways. Let's standardize on the lookup and keep the signature the same unless it makes things significantly more difficult for some case.

**Comment by nickoneill** [comment]
Yep - those sound like reasonable changes (ignoring the title issue mentioned by @oderby - let's do Senator / Rep. [Last Name] for names and then just he/she if necessary). Signature talk inline 👆

**Comment by nickoneill** [comment]
🏓

**Comment by bernardwang** [comment]
Hi @samhavens, any updates on this PR? If you don't want to make these changes I would be happy take the issue!

**Comment by samhavens** [comment]
@bernardwang I totally dropped the ball on this and would love your help. Thank you.

**Comment by nickoneill** [comment]
Closed in favor of #299

**Pull request closed without merging**